lib/ay_tests/media_builder.rb

```diff
@@ -0,0 +1,150 @@
+
+    # Constructor
+    #
+    # @param [Pathname] base_dir Set the base directory. By default it uses
+    #   AYTests.base_dir.
+    # @param [Pathname] yast_url YaST repository URL
+    # @param [Pathname] iso_url  Base ISO URL
+    # @param [String]   version  Distribution version (+sle12+, +sle12_sp1+, etc.)
+    def initialize(yast_url:, iso_url:, version:, base_dir: nil)
+      # Directories
+      @base_dir           = base_dir || AYTests.base_dir
+      @cache_dir          = base_dir.join("cache")
+      @local_packages_dir = base_dir.join("rpms", version)
+      @boot_dir           = base_dir.join("boot_#{version}")
+      @obs_pkg_list_path  = base_dir.join("build_iso", "#{version}.obs_packages")
+
+      # URLs
+      @yast_url = yast_url
+      @iso_url  = iso_url
+
+      # Misc data
+      @version = version
+    end
+
+    # Build a new ISO using packages from OBS
+    #
+    # * Do some cleanup
+    # * Download the original ISO (relying on AYTests::IsoRepo)
+    # * Fetch required packages
+    # * Fetch local packages
+    # * Create the final ISO
+    #
+    # @return [Boolean] true if image was successfully built; otherwise
+    #   return false.
+    #
+    # @see cleanup
+    # @see download_iso
+    # @see fetch_obs_packages
+    # @see fetch_local_packages
+    # @see build_iso
+    def build(output_path = nil)
+      cleanup
+      iso_path = download_iso
+      fetch_obs_packages
+      fetch_local_packages
+      build_iso(iso_path, output_path || AYTests.obs_iso_path)
+    end
+
+    # Clean-up cache directory
+    def cleanup
+      log.info("Cleaning up #{cache_dir}")
+      FileUtils.rm_r(cache_dir)
+    end
+
+    # Download the base ISO
+    #
+    # It relies on IsoRepo which will take care of downloading the
+    # ISO when it's needed.
+    #
+    # @return [Pathname] Path to downloaded ISO.
+    #
+    # @see AYTests::IsoRepo.get
+    def download_iso
+      IsoRepo.get(iso_url)
+    end
+
+    # Fetch OBS packages
+    #
+    # Method #obs_pkg_list_path will point to the list of packages to be downloaded.
+    # Moreover, some extra packages are downloaded (grub2, libzypp, libsolv-tools and zypper).
+    def fetch_obs_packages
+      puts "\n**** Fetching all required packages ****"
+      system "zypper --root #{cache_dir} ar --no-gpgcheck #{yast_url} download-packages"
+      system "xargs -a #{obs_pkg_list_path} zypper --root #{cache_dir} --pkg-cache-dir=#{cache_dir} download"
+
+      puts "\n**** Fetching latest grub2 and libzypp packages ****"
+      system "zypper --root #{cache_dir} rr download-packages"
+      system "zypper --root #{cache_dir} ar --no-gpgcheck http://download.suse.de/ibs/SUSE:/SLE-12-SP1:/GA/standard/ download-packages"
+      system "zypper --root #{cache_dir} --pkg-cache-dir=#{cache_dir} download grub2-2.02~beta2"
+      system "zypper --root #{cache_dir} --pkg-cache-dir=#{cache_dir} download grub2-i386-pc-2.02~beta2"
```

Comment:
Yes, we need to move more things to configuration file (`definitions.yml`). But as a first step, I left this part as it was originally. But sure it should be moved.